File: packages/sdk/tests/helpers/conformance-verify.ts
import { mkdtemp, readFile, rm } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { isDeepStrictEqual, types as utilTypes } from 'node:util';
import { initialize, resetCaptureForTests } from '../../src/v1/runtime.js';
import { validateArtifact } from '../../src/v1/validation.js';
import type { CaptureSource, SourceRecord, SourceSink } from '../../src/v1/types.js';

const CASES = ['lifecycle', 'stream', 'error', 'unknown', 'rejection', 'shutdown'] as const;
const SECRET = 'semantic-layer-conformance-secret-value';
const primordialReflectApply = Reflect.apply;
const primordialSymbolKeyFor = Symbol.keyFor;

type AdapterModule = { createSource(input: { subject: unknown }): CaptureSource };
type DriverModule = {
  createSubject(): unknown | Promise<unknown>;
  expectations: Record<string, unknown>;
} & Record<string, unknown>;

export type ConformanceReport = { valid: boolean; cases: number; issues: string[] };

export async function verifyAdapterConformance(
  adapter: AdapterModule,
  driver: DriverModule,
): Promise<ConformanceReport> {
  const issues: string[] = [];
  const output = await mkdtemp(join(tmpdir(), 'semantic-layer-adapter-'));
  try {
    const controlSubject = await driver.createSubject();
    const runners = new Map<typeof CASES[number], (...args: unknown[]) => unknown>();
    const controls = new Map<typeof CASES[number], CaseOutcome>();
    for (const name of CASES) {
      const run = driver[name];
      if (typeof run !== 'function' || !(name in driver.expectations)) {
        issues.push(`CASE_CONTRACT_MISSING:${name}`);
        continue;
      }
      const caseRunner = run as (...args: unknown[]) => unknown;
      runners.set(name, caseRunner);
      controls.set(name, await observeCase(caseRunner, driver, controlSubject));
    }
    const subject = await driver.createSubject();
    const capture = initialize({ output, serviceName: 'adapter-conformance', secretValues: [SECRET] });
    const source = adapter.createSource({ subject });
    const sourceName = source.metadata.name;
    if (typeof sourceName !== 'string' || sourceName.length === 0) {
      issues.push('SOURCE_METADATA_NAME_MISSING');
    }
    capture.installSource(source);
    let probeSink: SourceSink | undefined;
    capture.installSource({
      metadata: {
        name: 'conformance:source-input-probe', seam: 'conformance.runtime-probe',
        identityDomain: 'conformance.runtime-probe', coverage: [],
      },
      install(sink) {
        probeSink = sink;
        return { deactivate() {}, drain() {} };
      },
    });
    const probeOpened = probeSink?.openTrace({ name: 'conformance.source-input-probe' });
    if (!probeOpened || !probeOpened.accepted) {
      issues.push('SOURCE_INPUT_PROBE_OPEN_REJECTED');
    } else {
      const identitySnapshot = { ...probeOpened.identity };
      let immutable = Object.isFrozen(probeOpened.identity);
      try {
        for (const key of ['runId', 'traceId', 'operationId'] as const) {
          if (Reflect.set(probeOpened.identity, key, `tampered-${key}`)) immutable = false;
        }
      } catch {
        immutable = false;
      }
      if (!immutable || !isDeepStrictEqual(probeOpened.identity, identitySnapshot)) {
        issues.push('SOURCE_TRACE_IDENTITY_MUTABLE');
      }
      const forgedIdentity = probeSink!.record({
        kind: 'log', phase: 'event', name: 'conformance.forged-identity',
        trace: { ...probeOpened.identity, traceId: 'trace_conformance_forged' }, native: null,
      });
      if (forgedIdentity.accepted || forgedIdentity.reason !== 'invalid_record') {
        issues.push('SOURCE_TRACE_IDENTITY_FORGERY_NOT_REJECTED');
      }
      await forgedIdentity.settled;
      const forgedLoss = probeSink!.record({
        kind: 'loss', phase: 'gap', name: 'conformance.forged-loss',
        trace: probeOpened.identity, native: null,
        loss: { reason: 'serialization_failure', stage: 'source', recoverable: false },
      } as unknown as SourceRecord);
      if (forgedLoss.accepted || forgedLoss.reason !== 'invalid_record') {
        issues.push('SOURCE_LOSS_KIND_NOT_REJECTED');
      }
      const probeEnded = probeSink!.record({
        kind: 'lifecycle', phase: 'end', name: 'conformance.source-input-probe',
        trace: probeOpened.identity, native: null,
      });
      if (!probeEnded.accepted) issues.push('SOURCE_INPUT_PROBE_END_REJECTED');
    }
    for (const name of CASES) {
      const caseRunner = runners.get(name);
      const control = controls.get(name);
      if (!caseRunner || !control) continue;
      const observed = await observeCase(caseRunner, driver, subject);
      if (control.status === 'unsupported' || observed.status === 'unsupported') {
        issues.push(`CASE_OUTCOME_UNSUPPORTED:${name}`);
        continue;
      }
      if (!isDeepStrictEqual(observed, control)) issues.push(`CASE_PARITY_MISMATCH:${name}`);
      if (observed.status === 'returned'
        && observed.protocol === undefined
        && !isDeepStrictEqual(observed.value, comparable(driver.expectations[name]))) {
        issues.push(`CASE_RETURN_MISMATCH:${name}`);
      }
    }
    const status = await capture.shutdown();
    const artifact = await validateArtifact(status.artifactPath);
    if (!artifact.valid) issues.push(...artifact.issues.map((issue) => `ARTIFACT:${issue}`));
    const text = await readFile(join(status.artifactPath, 'trace.jsonl'), 'utf8');
    const rows = text.trim().split('\n').filter(Boolean).map((line) => JSON.parse(line)) as Array<Record<string, any>>;
    const manifest = JSON.parse(
      await readFile(join(status.artifactPath, 'manifest.json'), 'utf8'),
    ) as { sources?: Array<{ id?: string; name?: string }> };
    const sourceId = manifest.sources?.find((source) => source.name === sourceName)?.id;
    if (!rows.some((row) => row.kind === 'loss'
      && row.data?.reason === 'source_rejection')) {
      issues.push('SOURCE_TRACE_IDENTITY_REJECTION_EVIDENCE_MISSING');
    }
    if (!rows.some((row) => row.kind === 'loss'
      && row.data?.reason === 'serialization_failure'
      && row.data?.path === '/event_kind')) {
      issues.push('SOURCE_LOSS_REJECTION_EVIDENCE_MISSING');
    }
    const sourceRows = rows.filter((row) => row.source === sourceId);
    if (sourceRows.length === 0) issues.push('SOURCE_EVIDENCE_MISSING');
    const starts = sourceRows.filter((row) => row.kind === 'run.start');
    const outcomes = sourceRows.filter((row) => row.kind === 'run.outcome');
    if (starts.length === 0) issues.push('LIFECYCLE_START_MISSING');
    if (!outcomes.some((row) => row.data?.status === 'completed')) issues.push('LIFECYCLE_END_MISSING');
    if (!sourceRows.some((row) => row.kind === 'message')) issues.push('STREAM_EVIDENCE_MISSING');
    if (!sourceRows.some((row) => row.kind === 'error')) issues.push('ERROR_EVIDENCE_MISSING');
    if (!sourceRows.some((row) => row.kind === 'loss'
      && row.data?.reason === 'unsupported_semantic_projection')) {
      issues.push('UNKNOWN_EVIDENCE_MISSING');
    }
    if (!outcomes.some((row) => row.data?.status === 'cancelled')) issues.push('ACTIVE_SHUTDOWN_MISSING');
    const startIds = new Set(starts.map((row) => row.id));
    if (outcomes.some((row) => !startIds.has(row.parent))) issues.push('SOURCE_PARENTAGE_INVALID');
    if (status.queue.pendingBytes !== 0 || status.queue.pendingControlBytes !== 0) {
      issues.push('ADMITTED_RECORDS_UNSETTLED');
    }
    if (!rows.some((row) => row.kind === 'loss'
      && ['credential_redaction', 'scrubber_failure_payload_omitted'].includes(row.data?.reason))) {
      issues.push('SECRET_REDACTION_LOSS_MISSING');
    }
    if (text.includes(SECRET)) issues.push('SECRET_PERSISTED');
    return { valid: issues.length === 0, cases: CASES.length, issues };
  } catch (error) {
    return {
      valid: false,
      cases: CASES.length,
      issues: [`ADAPTER_CONFORMANCE_FAILED:${error instanceof Error ? `${error.name}:${error.message}` : String(error)}`],
    };
  } finally {
    await resetCaptureForTests();
    await rm(output, { recursive: true, force: true });
  }
}

type CaseOutcome = {
  status: 'returned' | 'threw' | 'unsupported';
  value?: unknown;
  error?: unknown;
  protocol?: unknown;
};

async function observeCase(
  run: (...args: unknown[]) => unknown,
  receiver: unknown,
  subject: unknown,
): Promise<CaseOutcome> {
  let value: unknown;
  try {
    value = await Reflect.apply(run, receiver, [subject]);
  } catch (error) {
    try {
      return { status: 'threw', error: comparableError(error) };
    } catch (observationError) {
      return { status: 'unsupported', error: observationReason(observationError) };
    }
  }
  try {
    const iteratorFactory = dataMethod(value, Symbol.asyncIterator);
    if (iteratorFactory) {
      return { status: 'returned', protocol: await observeAsyncProtocol(value as object, iteratorFactory) };
    }
    return { status: 'returned', value: comparable(value) };
  } catch (error) {
    return { status: 'unsupported', error: observationReason(error) };
  }
}

async function observeAsyncProtocol(
  iterable: object,
  factory: (...args: unknown[]) => unknown,
): Promise<unknown> {
  const values: unknown[] = [];
  const iterator = Reflect.apply(factory, iterable, []) as object;
  const next = requiredDataMethod(iterator, 'next');
  for (let index = 0; index < 64; index += 1) {
    const part = await Reflect.apply(next, iterator, []) as IteratorResult<unknown>;
    values.push(comparable(part));
    if (part.done) break;
    if (index === 63) values.push({ bounded: true });
  }
  const returnSentinel = { conformance: 'return-sentinel' };
  const returnIterator = Reflect.apply(factory, iterable, []) as object;
  const returnMethod = dataMethod(returnIterator, 'return');
  const returned = returnMethod
    ? await settleProtocol(
      () => Promise.resolve(Reflect.apply(returnMethod, returnIterator, [returnSentinel])) as Promise<IteratorResult<unknown>>,
      returnSentinel,
    )
    : { supported: false };
  const throwSentinel = new Error('conformance-throw-sentinel');
  const throwIterator = Reflect.apply(factory, iterable, []) as object;
  const throwMethod = dataMethod(throwIterator, 'throw');
  const thrown = throwMethod
    ? await settleProtocol(
      () => Promise.resolve(Reflect.apply(throwMethod, throwIterator, [throwSentinel])) as Promise<IteratorResult<unknown>>,
      throwSentinel,
    )
    : { supported: false };
  return { values, returned, thrown };
}

async function settleProtocol(run: () => Promise<IteratorResult<unknown>>, sentinel: unknown): Promise<unknown> {
  let result: IteratorResult<unknown>;
  try {
    result = await run();
  } catch (error) {
    return { supported: true, status: 'threw', error: comparableError(error), sentinel: error === sentinel };
  }
  return { supported: true, status: 'returned', result: comparable(result), sentinel: result.value === sentinel };
}

function comparable(value: unknown, seen = new WeakSet<object>()): unknown {
  if (value === null || value === undefined || ['string', 'number', 'boolean'].includes(typeof value)) {
    return value;
  }
  if (typeof value === 'bigint') return { bigint: value.toString() };
  if (typeof value === 'symbol') return symbolSnapshot(value);
  if ((typeof value !== 'object' && typeof value !== 'function') || utilTypes.isProxy(value)) {
    throw new ConformanceObservationError('unsupported or proxied return value');
  }
  if (typeof value === 'function') throw new ConformanceObservationError('function return value');
  if (seen.has(value)) return { circular: true };
  seen.add(value);
  if (utilTypes.isMap(value)) {
    return { prototype: prototypeIdentity(Object.getPrototypeOf(value)), map: [...Map.prototype.entries.call(value) as Iterable<[unknown, unknown]>]
      .map(([key, item]) => [comparable(key, seen), comparable(item, seen)]) };
  }
  if (utilTypes.isSet(value)) {
    return { prototype: prototypeIdentity(Object.getPrototypeOf(value)), set: [...Set.prototype.values.call(value) as Iterable<unknown>]
      .map((item) => comparable(item, seen)) };
  }
  const descriptors = Object.getOwnPropertyDescriptors(value);
  const keys = Reflect.ownKeys(descriptors)
    .filter((key) => !(utilTypes.isNativeError(value) && key === 'stack'))
    .sort((left, right) => propertyLabel(left).localeCompare(propertyLabel(right)));
  const prototype = Object.getPrototypeOf(value) as object | null;
  if (keys.length === 0 && prototype !== Object.prototype && prototype !== null) {
    throw new ConformanceObservationError('opaque return value');
  }
  return {
    prototype: prototypeIdentity(prototype),
    object: keys.map((key) => {
      const descriptor = descriptors[key as keyof typeof descriptors]!;
      if (!('value' in descriptor)) {
        throw new ConformanceObservationError(`accessor return property: ${propertyLabel(key)}`);
      }
      return {
        key: propertySnapshot(key), value: comparable(descriptor.value, seen),
        enumerable: descriptor.enumerable, configurable: descriptor.configurable,
        writable: descriptor.writable,
      };
    }),
  };
}

function comparableError(error: unknown): unknown {
  return comparable(error);
}

class ConformanceObservationError extends Error {}

const prototypeIdentities = new WeakMap<object, number>();
let nextPrototypeIdentity = 0;

function prototypeIdentity(prototype: object | null): number {
  if (prototype === null) return 0;
  const existing = prototypeIdentities.get(prototype);
  if (existing !== undefined) return existing;
  const assigned = ++nextPrototypeIdentity;
  prototypeIdentities.set(prototype, assigned);
  return assigned;
}

function observationReason(error: unknown): string {
  return error instanceof ConformanceObservationError ? error.message : 'observation failed';
}

function propertyLabel(key: PropertyKey): string {
  if (typeof key !== 'symbol') return `string:${key}`;
  const snapshot = symbolSnapshot(key);
  return snapshot.registry === 'global'
    ? `symbol:global:${snapshot.key}`
    : 'symbol:local';
}

function symbolSnapshot(value: symbol):
  | { symbol: true; registry: 'global'; key: string }
  | { symbol: true; registry: 'local'; identity: symbol } {
  const key = primordialReflectApply(
    primordialSymbolKeyFor,
    Symbol,
    [value],
  ) as string | undefined;
  if (key !== undefined) return { symbol: true, registry: 'global', key };
  return { symbol: true, registry: 'local', identity: value };
}

function propertySnapshot(key: PropertyKey): unknown {
  return typeof key === 'symbol'
    ? symbolSnapshot(key)
    : { property: 'string', value: key };
}

function dataMethod(value: unknown, key: PropertyKey): ((...args: unknown[]) => unknown) | undefined {
  if (value === null || (typeof value !== 'object' && typeof value !== 'function')) return undefined;
  if (utilTypes.isProxy(value)) throw new ConformanceObservationError('proxied protocol surface');
  let current: object | null = value;
  while (current !== null) {
    if (utilTypes.isProxy(current)) throw new ConformanceObservationError('proxied protocol prototype');
    const descriptor = Object.getOwnPropertyDescriptor(current, key);
    if (descriptor) {
      if (!('value' in descriptor)) {
        throw new ConformanceObservationError(`accessor protocol property: ${propertyLabel(key)}`);
      }
      if (descriptor.value === undefined) return undefined;
      if (typeof descriptor.value !== 'function') {
        throw new ConformanceObservationError(`non-callable protocol property: ${propertyLabel(key)}`);
      }
      return descriptor.value as (...args: unknown[]) => unknown;
    }
    current = Object.getPrototypeOf(current) as object | null;
  }
  return undefined;
}

function requiredDataMethod(value: unknown, key: PropertyKey): (...args: unknown[]) => unknown {
  const method = dataMethod(value, key);
  if (!method) throw new ConformanceObservationError(`missing protocol property: ${propertyLabel(key)}`);
  return method;
}

File: packages/python/tests/conformance_verify.py
"""Repository-only verifier for source conformance."""

from __future__ import annotations

import inspect
import json
import struct
import tempfile
from collections.abc import Mapping
from pathlib import Path
from typing import Any

from semantic_layer.capture_v1 import CaptureSource, initialize, reset_capture_for_tests
from semantic_layer.validation import validate_artifact

_CASES = ("lifecycle", "stream", "error", "unknown", "rejection", "shutdown")
_SECRET = "semantic-layer-conformance-secret-value"


async def verify_conformance(adapter_module: Any, driver_module: Any) -> dict[str, Any]:
    issues: list[str] = []
    try:
        with tempfile.TemporaryDirectory(prefix="semantic-layer-adapter-") as directory:
            try:
                control_subject = await _await_maybe(driver_module.create_subject())
                runners: dict[str, Any] = {}
                controls: dict[str, dict[str, Any]] = {}
                for case in _CASES:
                    run = getattr(driver_module, case, None)
                    expectations = getattr(driver_module, "expectations", {})
                    if not callable(run) or case not in expectations:
                        issues.append(f"CASE_CONTRACT_MISSING:{case}")
                        continue
                    runners[case] = run
                    controls[case] = await _observe_case(run, control_subject)
                subject = await _await_maybe(driver_module.create_subject())
                capture = initialize(
                    output=directory,
                    service_name="adapter-conformance",
                    secret_values=[_SECRET],
                )
                source = adapter_module.create_source(subject=subject)
                source_name = source.metadata.get("name")
                if not isinstance(source_name, str) or not source_name:
                    issues.append("SOURCE_METADATA_NAME_MISSING")
                capture.install_source(source)
                probe_sink: Any = None

                class _ProbeSource(CaptureSource):
                    metadata = {
                        "name": "conformance:source-input-probe",
                        "seam": "conformance.runtime-probe",
                        "identity_domain": "conformance.runtime-probe",
                        "coverage": [],
                    }

                    def install(self, sink: Any) -> Any:
                        nonlocal probe_sink
                        probe_sink = sink

                        class _Lifecycle:
                            def deactivate(self) -> None:
                                pass

                            def drain(self) -> None:
                                pass

                        return _Lifecycle()

                capture.install_source(_ProbeSource())
                probe_opened = probe_sink.open_trace(
                    {"name": "conformance.source-input-probe"}
                )
                if not probe_opened.accepted:
                    issues.append("SOURCE_INPUT_PROBE_OPEN_REJECTED")
                else:
                    identity = probe_opened.identity
                    identity_snapshot = dict(identity)
                    immutable = isinstance(identity, Mapping)
                    for key in ("run_id", "trace_id", "operation_id"):
                        try:
                            identity[key] = f"tampered-{key}"
                            immutable = False
                        except TypeError:
                            pass
                        except BaseException:
                            immutable = False
                    if not immutable or dict(identity) != identity_snapshot:
                        issues.append("SOURCE_TRACE_IDENTITY_MUTABLE")
                    forged_identity = probe_sink.record(
                        {
                            "kind": "log",
                            "phase": "event",
                            "name": "conformance.forged-identity",
                            "trace": {
                                **dict(probe_opened.identity),
                                "trace_id": "trace_conformance_forged",
                            },
                            "native": None,
                        }
                    )
                    if (
                        forged_identity.accepted
                        or forged_identity.reason != "invalid_record"
                    ):
                        issues.append("SOURCE_TRACE_IDENTITY_FORGERY_NOT_REJECTED")
                    await forged_identity.settled
                    forged_loss = probe_sink.record(
                        {
                            "kind": "loss",
                            "phase": "gap",
                            "name": "conformance.forged-loss",
                            "trace": probe_opened.identity,
                            "native": None,
                            "loss": {
                                "reason": "serialization_failure",
                                "stage": "source",
                                "recoverable": False,
                            },
                        }
                    )
                    if forged_loss.accepted or forged_loss.reason != "invalid_record":
                        issues.append("SOURCE_LOSS_KIND_NOT_REJECTED")
                    probe_ended = probe_sink.record(
                        {
                            "kind": "lifecycle",
                            "phase": "end",
                            "name": "conformance.source-input-probe",
                            "trace": probe_opened.identity,
                            "native": None,
                            "semantic": {
                                "type": "workflow.run",
                                "status": "succeeded",
                            },
                        }
                    )
                    if not probe_ended.accepted:
                        issues.append("SOURCE_INPUT_PROBE_END_REJECTED")
                for case in _CASES:
                    run = runners.get(case)
                    control = controls.get(case)
                    if run is None or control is None:
                        continue
                    expectations = getattr(driver_module, "expectations", {})
                    observed = await _observe_case(run, subject)
                    if (
                        control.get("status") == "unsupported"
                        or observed.get("status") == "unsupported"
                    ):
                        issues.append(f"CASE_OUTCOME_UNSUPPORTED:{case}")
                        continue
                    if observed != control:
                        issues.append(f"CASE_PARITY_MISMATCH:{case}")
                    if (
                        observed.get("status") == "returned"
                        and "protocol" not in observed
                        and observed.get("value") != _comparable(expectations[case])
                    ):
                        issues.append(f"CASE_RETURN_MISMATCH:{case}")
                status = capture.shutdown()
                artifact = validate_artifact(status.artifact_path)
                issues.extend(f"ARTIFACT:{issue}" for issue in artifact.issues)
                text = (Path(status.artifact_path) / "trace.jsonl").read_text()
                rows = [json.loads(line) for line in text.splitlines()]
                manifest = json.loads(
                    (Path(status.artifact_path) / "manifest.json").read_text()
                )
                source_id = next(
                    (
                        source["id"]
                        for source in manifest["sources"]
                        if source["name"] == source_name
                    ),
                    None,
                )
                if not any(
                    row.get("kind") == "loss"
                    and row.get("data", {}).get("reason") == "source_rejection"
                    and row.get("data", {}).get("path") == "/trace"
                    for row in rows
                ):
                    issues.append("SOURCE_TRACE_IDENTITY_REJECTION_EVIDENCE_MISSING")
                if not any(
                    row.get("kind") == "loss"
                    and row.get("data", {}).get("reason")
                    == "serialization_failure"
                    and row.get("data", {}).get("path") == "/event_kind"
                    for row in rows
                ):
                    issues.append("SOURCE_LOSS_REJECTION_EVIDENCE_MISSING")
                source_rows = [
                    row for row in rows if row.get("source") == source_id
                ]
                if not source_rows:
                    issues.append("SOURCE_EVIDENCE_MISSING")
                if not any(
                    row["kind"] == "run.start"
                    for row in source_rows
                ):
                    issues.append("LIFECYCLE_START_MISSING")
                if not any(
                    row["kind"] == "run.outcome"
                    and row.get("data", {}).get("status") == "completed"
                    for row in source_rows
                ):
                    issues.append("LIFECYCLE_END_MISSING")
                if not any(
                    row["kind"] == "state"
                    and row.get("data", {}).get("type") == "state.stream_observed"
                    for row in source_rows
                ):
                    issues.append("STREAM_EVIDENCE_MISSING")
                if not any(row["kind"] == "error" for row in source_rows):
                    issues.append("ERROR_EVIDENCE_MISSING")
                if not any(
                    row["kind"] == "state"
                    and row.get("data", {}).get("type") == "state.unknown_observed"
                    for row in source_rows
                ):
                    issues.append("UNKNOWN_EVIDENCE_MISSING")
                if not any(
                    row["kind"] == "run.outcome"
                    and row.get("data", {}).get("status") == "cancelled"
                    for row in source_rows
                ):
                    issues.append("ACTIVE_SHUTDOWN_MISSING")
                if (
                    status.pending_bytes != 0
                    or status.pending_control_bytes != 0
                ):
                    issues.append("ADMITTED_RECORDS_UNSETTLED")
                if not any(
                    row.get("kind") == "loss"
                    and row.get("data", {}).get("reason") == "credential_redaction"
                    for row in rows
                ):
                    issues.append("SECRET_REDACTION_LOSS_MISSING")
                if _SECRET in text:
                    issues.append("SECRET_PERSISTED")
            finally:
                reset_capture_for_tests()
    except Exception as error:
        issues = [f"ADAPTER_CONFORMANCE_FAILED:{type(error).__name__}:{error}"]
    return {"valid": not issues, "cases": len(_CASES), "issues": issues}


async def _observe_case(run: Any, subject: Any) -> dict[str, Any]:
    try:
        value = await _await_maybe(run(subject))
    except Exception as error:
        try:
            return {"status": "threw", "error": _error_value(error)}
        except _ConformanceObservationError as observation_error:
            return {"status": "unsupported", "error": str(observation_error)}
    try:
        iterator_factory = _special_method(value, "__aiter__")
        if iterator_factory is not None:
            return {
                "status": "returned",
                "protocol": await _observe_async_protocol(value, iterator_factory),
            }
        return {"status": "returned", "value": _comparable(value)}
    except _ConformanceObservationError as error:
        return {"status": "unsupported", "error": str(error)}


async def _observe_async_protocol(iterable: Any, iterator_factory: Any) -> dict[str, Any]:
    values: list[Any] = []
    iterator = iterator_factory(iterable)
    next_method = _required_special_method(iterator, "__anext__")
    for index in range(64):
        try:
            value = await next_method(iterator)
            values.append({"done": False, "value": _comparable(value)})
        except StopAsyncIteration:
            values.append({"done": True})
            break
        if index == 63:
            values.append({"bounded": True})
    return_iterator = iterator_factory(iterable)
    close = _special_method(return_iterator, "aclose")
    returned = (
        await _settle_protocol(close, return_iterator, None)
        if close is not None
        else {"supported": False}
    )
    throw_iterator = iterator_factory(iterable)
    throw = _special_method(throw_iterator, "athrow")
    sentinel = RuntimeError("conformance-throw-sentinel")
    thrown = (
        await _settle_protocol(throw, throw_iterator, sentinel, sentinel)
        if throw is not None
        else {"supported": False}
    )
    return {"values": values, "returned": returned, "thrown": thrown}


async def _settle_protocol(
    operation: Any, receiver: Any, argument: Any, sentinel: Any = None
) -> dict[str, Any]:
    try:
        result = (
            await operation(receiver, argument)
            if argument is not None
            else await operation(receiver)
        )
    except Exception as error:
        return {
            "supported": True,
            "status": "threw",
            "error": _error_value(error),
            "sentinel": error is sentinel,
        }
    return {
        "supported": True,
        "status": "returned",
        "result": _comparable(result),
        "sentinel": result is sentinel,
    }


async def _await_maybe(value: Any) -> Any:
    return await value if inspect.isawaitable(value) else value


def _error_value(error: BaseException) -> dict[str, Any]:
    args = object.__getattribute__(error, "args")
    cause = object.__getattribute__(error, "__cause__")
    return {
        "type": type(error).__name__,
        "args": _comparable(args),
        "cause": _comparable(cause),
    }


def _comparable(value: Any, seen: set[int] | None = None) -> Any:
    value_type = type(value)
    if value is None:
        return {"type": "none"}
    if value_type is bool:
        return {"type": "bool", "value": value}
    if value_type is int:
        return {"type": "int", "value": value}
    if value_type is str:
        return {"type": "str", "value": value}
    if value_type is float:
        return {"type": "float", "ieee754": struct.pack(">d", value).hex()}
    if isinstance(value, (str, int, float, bool)):
        raise _ConformanceObservationError("primitive subclass return shape")
    if isinstance(value, BaseException):
        return _error_value(value)
    identities = seen if seen is not None else set()
    identity = id(value)
    if identity in identities:
        return {"circular": True}
    identities.add(identity)
    if type(value) in (list, tuple):
        return {
            "sequence": type(value).__name__,
            "values": [_comparable(item, identities) for item in value],
        }
    if type(value) is dict:
        return {
            "map": [
                [_comparable(key, identities), _comparable(item, identities)]
                for key, item in dict.items(value)
            ]
        }
    if type(value) is set:
        items = [_comparable(item, identities) for item in set.__iter__(value)]
        return {"set": sorted(items, key=lambda item: json.dumps(item, sort_keys=True))}
    for cls in type.__getattribute__(value_type, "__mro__"):
        namespace = type.__getattribute__(cls, "__dict__")
        if any(isinstance(descriptor, property) for descriptor in namespace.values()):
            raise _ConformanceObservationError("property return shape")
    try:
        attributes = object.__getattribute__(value, "__dict__")
    except AttributeError as error:
        raise _ConformanceObservationError("opaque return shape") from error
    if type(attributes) is not dict or not attributes:
        raise _ConformanceObservationError("unsupported empty return shape")
    return {
        "object": f"{value_type.__module__}.{value_type.__qualname__}",
        "attributes": _comparable(attributes, identities),
    }


class _ConformanceObservationError(Exception):
    pass


def _special_method(value: Any, name: str) -> Any:
    if value is None:
        return None
    value_type = type(value)
    for cls in type.__getattribute__(value_type, "__mro__"):
        namespace = type.__getattribute__(cls, "__dict__")
        if name not in namespace:
            continue
        descriptor = namespace[name]
        if isinstance(descriptor, property):
            raise _ConformanceObservationError(f"property protocol surface: {name}")
        if not callable(descriptor):
            raise _ConformanceObservationError(f"non-callable protocol surface: {name}")
        return descriptor
    return None


def _required_special_method(value: Any, name: str) -> Any:
    method = _special_method(value, name)
    if method is None:
        raise _ConformanceObservationError(f"missing protocol surface: {name}")
    return method
